wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -351,24 +351,31 @@ public void stopWorkspace(String workspaceId, Map<String, String> options)
   }
 
   /** Asynchronously starts given workspace. */
-  private void startAsync(WorkspaceImpl workspace, String envName, Map<String, String> options)
+  private void startAsync(
+      WorkspaceImpl workspace, @Nullable String envName, Map<String, String> options)
       throws ConflictException, NotFoundException, ServerException {
-    if (envName != null && !workspace.getConfig().getEnvironments().containsKey(envName)) {
-      throw new NotFoundException(
-          format(
-              "Workspace '%s:%s' doesn't contain environment '%s'",
-              workspace.getNamespace(), workspace.getConfig().getName(), envName));
+    // Sidecar-based workspaces allowed not to have any environments
+    String env = null;
+    if (!SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes())
```

Comment:
What about:
```
boolan nonSidecarWorkspace = !SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes());
boolean envListNonEmpty = workspace.getConfig().getEnvironments().isEmpty();
if (nonSidecarWorkspace || envName != null || envListNonEmpty) {
// check env
}
```
Which means that body of the if clause would be executed if any of next conditions is true:
- it is Che 6 WS
- default env is set 
- environments list is empty 

WDYT?

Reply:
Well, I investigated this code and discovered an interesting thing for me:
`!workspace.getConfig().getEnvironments().isEmpty()` means that if there is one environment then this environment must be configured as default. Am I right?

WorkspaceValidator checks that `DefaultEnv` references existing environment, then maybe it would be better to rewrite check like 
```
if (nonSidecarWorkspace || envName != null || workspace.getConfig().getDefaultEnv) {
}
```

Then it may be rewritten even like in the following way:
```
   if (envName == null) {
      envName = workspace.getConfig().getDefaultEnv()
    }

    // Sidecar-based workspaces allowed not to have any environments
    if (!SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes())
        || envName != null) {
      if (envName == null) {
          throw new NotFoundException("Missing environment to start");
      }
      if (!workspace.getConfig().getEnvironments().containsKey(envName)) {
        throw new NotFoundException(
            format(
                "Workspace '%s:%s' doesn't contain environment '%s'",
                workspace.getNamespace(), workspace.getConfig().getName(), envName));
      }

      // validate environment in advance
      try {
        runtimes.validate(workspace.getConfig().getEnvironments().get(envName));
      } catch (InfrastructureException | ValidationException e) {
        throw new ServerException(e);
      }
    }
```

Reply:
More more way:
```java
  /** Asynchronously starts given workspace. */
  private void startAsync(
      WorkspaceImpl workspace, @Nullable String envName, Map<String, String> options)
      throws ConflictException, NotFoundException, ServerException {
    String env = getValidEnvironmentName(workspace, envName);
    workspace.getAttributes().put(UPDATED_ATTRIBUTE_NAME, Long.toString(currentTimeMillis()));
    workspaceDao.update(workspace);
    runtimes
        .startAsync(workspace, env, firstNonNull(options, Collections.emptyMap()))
        .thenAccept(aVoid -> handleStartupSuccess(workspace))
        .exceptionally(
            ex -> {
              if (workspace.isTemporary()) {
                removeWorkspaceQuietly(workspace);
              } else {
                handleStartupError(workspace, ex.getCause());
              }
              return null;
            });
  }

  private String getValidEnvironmentName(WorkspaceImpl workspace, String envName)
      throws NotFoundException, ServerException {
      if (envName != null && !workspace.getConfig().getEnvironments().containsKey(envName)) {
        throw new NotFoundException(
            format(
                "Workspace '%s:%s' doesn't contain environment '%s'",
                workspace.getNamespace(), workspace.getConfig().getName(), envName));
      }
    
      envName = firstNonNull(envName, workspace.getConfig().getDefaultEnv());

      if (envName == null && SidecarToolingWorkspaceUtil.isSidecarBasedWorkspace(workspace.getConfig().getAttributes()) {
          // Sidecar-based workspaces allowed not to have any environments
          return null; 
      }

      // validate environment in advance
      try {
        runtimes.validate(workspace.getConfig().getEnvironments().get(envName));
      } catch (InfrastructureException | ValidationException e) {
        throw new ServerException(e);
      }
  }
```

Reply:
> Well, I investigated this code and discovered an interesting thing for me:
!workspace.getConfig().getEnvironments().isEmpty() means that if there is one environment then this environment must be configured as default. Am I right?

absolutely

I like the last approach you suggested, will apply it. Thanks!